### PR TITLE
fix: upgrade to JCS3

### LINKED ISSUE #293: Cache Misses In CentralAnalyzer When Gradle Daemon Is Reused
### Bug Description

When `CentralAnalyzer` is enabled caching is meant to work by default. However, repeated executions of ODC result in cache misses. This makes a noticeable difference for projects with many dependencies because each dependency adds about 1 second of runtime.

This issue focuses on caching for the `CentalAnalyzer` but affects every analyzer using the `DataCacheFactory`, i.e. implementations of `AbstractNpmAnalyzer` are concerned as well.

Tests were conducted using version 7.3.2 of the ODC plugin and Gradle 7.1.

#### Cause Description

The `DataCacheFactory` uses JCS for the underlying caching mechanism. The `CompositeCacheManager` requested inside `JCS#getInstance( ...)` is a singleton. After creating its instance it is configured to use an `IndexedDiskCache` (also called auxiliary cache) which is divided into regions POM, CENTRAL and NODEAUDIT. Since ODC enables caches from the `CacheExtension` by default these sub-caches create files in the data directory (as set via `DataExtension`) which are populated while dependency check is running. When ODC finishes it calls `JCS#shutdown()` that cleans up memory. As a consequence cache values get disposed and aux caches are removed from the map of all caches (`CompositeCacheManager#caches`). However, `CompositeCacheManager#auxiliaryCaches` is left untouched which means that the aux caches that were just disposed are still referenced inside the cache manager which is not invalidated as well.

Because Gradle uses the daemon by default since version 3 this has implications for using [static state in plugins](https://github.com/gradle/gradle/issues/6818). If the same daemon is used again for repeated executions of ODC the JCS cache manager (still alive) is also reused. While trying to configure it again the aux cache map containing the regions mentioned above can be retrieved without having to recreate its entries in an `AuxiliaryCacheFactory`. As a result the aux caches in memory are not connected to their respective files anymore which effectively leads to empty caches being offered by JCS. Therefore, further invocations of `CentralAnalyzer` will start dowloading POMs again that users have cached on disk.

#### Repro

See [mwe.zip](https://github.com/dependency-check/dependency-check-gradle/files/10102430/mwe.zip) attached for reference.

1. Delete `<USER_HOME>\.gradle\dependency-check-data\7.0\cache`
2. Switch to `myproject`
3. `gradlew :dependencyCheckAggregate --info`

        Instance is null, returning unconfigured instance

        thread_pool.default PoolConfiguration = useBoundary = [false] boundarySize = [2000] maximumPoolSize = [150] minimumPoolSize = [4] keepAliveTime = [300000] whenBlockedPolicy = [RUN] startUpSize = [4]
        Setting default auxiliaries to ODC
        setting defaultCompositeCacheAttributes to [ useLateral = true, useRemote = true, useDisk = true, maxObjs = 0, maxSpoolPerRun = -1, diskUsagePattern = UPDATE, spoolChunkSize = 2 ]
        setting defaultElementAttributes to [ IS_LATERAL = false, IS_SPOOL = true, IS_REMOTE = false, IS_ETERNAL = false, MaxLifeSeconds = 86400, IdleTime = 1800, CreateTime = 1669485286945, LastAccessTime = 1669485286945, getTimeToLiveSeconds() = 86399, createTime = 1669485286945 ]
        
        [...]

        initialized MemoryCache for POM
        Constructed cache with name [POM] and cache attributes [ useLateral = true, useRemote = true, useDisk = true, maxObjs = 0, maxSpoolPerRun = -1, diskUsagePattern = UPDATE, spoolChunkSize = 2 ]
        No cache event logger defined for auxiliary [jcs.auxiliary.ODC]
        Using standard serializer [org.apache.commons.jcs.utils.serialization.StandardSerializer@60a7c42f] for auxiliary [jcs.auxiliary.ODC]
        Region [POM] Cache file root directory: <USER_HOME>\.gradle\dependency-check-data\7.0\cache
        Region [POM] Set maxKeySize to: '1000000'
        Region [POM] Indexed Disk Cache is alive.

        [...]

        Parsed regions [CENTRAL, POM, NODEAUDIT]
        Finished configuration in 36 ms.

    - notice the creation of a brand new cache manager in the 1st line and the cache "alive" status
    - data directory is created with empty `.key` and `.data` files
    - `CentralAnalyzer` downloads POMs whose model information is saved to the `.data` files (...this takes some time which is OK for initially building the cache)

    

        Finished Central Analyzer (98 seconds)

    - finally, when ODC ends the `.key` files get filled:

    

        Element event queue destroyed: org.apache.commons.jcs.engine.control.event.ElementEventQueue@3d1f6bf0

        [...]
        
        No longer waiting for event queue to finish: Pooled Cache Event Queue
        Working = true
        Alive = false
        Empty = true
        Queue Size = 0
        Queue Capacity = 2147483647
        Pool Size = 0
        Maximum Pool Size = 150
        Region [POM] Saving keys to: POM, key count: 105
        Region [POM] Finished saving keys.
        Region [POM] Shutdown complete.
        
        [...]
        
        In DISPOSE, [POM] fromRemote [false]
        In DISPOSE, [POM] auxiliary [POM]
        In DISPOSE, [POM] put 0 into auxiliary POM
        In dispose, destroying event queue.
        In DISPOSE, [POM] disposing of memory cache.
        Memory Cache dispose called.

4. Repeat step 3 to validate caching

        thread_pool.default PoolConfiguration = useBoundary = [false] boundarySize = [2000] maximumPoolSize = [150] minimumPoolSize = [4] keepAliveTime = [300000] whenBlockedPolicy = [RUN] startUpSize = [4]
        
        Setting default auxiliaries to ODC
        setting defaultCompositeCacheAttributes to [ useLateral = true, useRemote = true, useDisk = true, maxObjs = 0, maxSpoolPerRun = -1, diskUsagePattern = UPDATE, spoolChunkSize = 2 ]
        setting defaultElementAttributes to [ IS_LATERAL = false, IS_SPOOL = true, IS_REMOTE = false, IS_ETERNAL = false, MaxLifeSeconds = 86400, IdleTime = 1800, CreateTime = 1669520238206, LastAccessTime = 1669520238206, getTimeToLiveSeconds() = 86399, createTime = 1669520238206 ]
        
        [...]

        initialized MemoryCache for POM
        Constructed cache with name [POM] and cache attributes [ useLateral = true, useRemote = true, useDisk = true, maxObjs = 0, maxSpoolPerRun = -1, diskUsagePattern = UPDATE, spoolChunkSize = 2 ]
        
        [...]
        
        Parsed regions [CENTRAL, POM, NODEAUDIT]
        Finished configuration in 1 ms.

    - `IndexedDiskCache` isn't mentioned
    - aux cache regions are no longer alive (`CacheStatus#DISPOSED`) as communicated at DEBUG:

    

        [...]

        [org.apache.commons.jcs.engine.control.CompositeCache] get: key = https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom, localOnly = false
        [org.apache.commons.jcs.engine.memory.AbstractDoubleLinkedListMemoryCache] POM: getting item for key https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.apache.commons.jcs.engine.memory.AbstractDoubleLinkedListMemoryCache] POM: LRUMemoryCache miss for https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        
        [...]

        [org.apache.commons.jcs.engine.control.CompositeCache] Attempting to get from aux [POM] which is of type: DISK_CACHE
        [org.apache.commons.jcs.auxiliary.disk.AbstractDiskCache] get was called, but the disk cache is not alive.
        [org.apache.commons.jcs.engine.control.CompositeCache] Got CacheElement: null
        [org.apache.commons.jcs.engine.control.CompositeCache] POM - Miss
        [org.owasp.dependencycheck.analyzer.CentralAnalyzer] Downloading https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom

        [...]

    - `CentralAnalyzer` takes as much time as before (around 100 seconds)

5. `gradlew :dependencyCheckAggregate --info --no-daemon`
    - run this outside an IDE that uses the [Gradle Tooling API property](https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:tools_and_ides)

    

        Instance is null, returning unconfigured instance
        
        thread_pool.default PoolConfiguration = useBoundary = [false] boundarySize = [2000] maximumPoolSize = [150] minimumPoolSize = [4] keepAliveTime = [300000] whenBlockedPolicy = [RUN] startUpSize = [4]
        Setting default auxiliaries to ODC
        setting defaultCompositeCacheAttributes to [ useLateral = true, useRemote = true, useDisk = true, maxObjs = 0, maxSpoolPerRun = -1, diskUsagePattern = UPDATE, spoolChunkSize = 2 ]
        setting defaultElementAttributes to [ IS_LATERAL = false, IS_SPOOL = true, IS_REMOTE = false, IS_ETERNAL = false, MaxLifeSeconds = 86400, IdleTime = 1800, CreateTime = 1669519678636, LastAccessTime = 1669519678636, getTimeToLiveSeconds() = 86399, createTime = 1669519678636 ]
        
        [...]
        
        initialized MemoryCache for POM
        Constructed cache with name [POM] and cache attributes [ useLateral = true, useRemote = true, useDisk = true, maxObjs = 0, maxSpoolPerRun = -1, diskUsagePattern = UPDATE, spoolChunkSize = 2 ]
        No cache event logger defined for auxiliary [jcs.auxiliary.ODC]
        Using standard serializer [org.apache.commons.jcs.utils.serialization.StandardSerializer@7b49312a] for auxiliary [jcs.auxiliary.ODC]
        Region [POM] Cache file root directory: <USER_HOME>\.gradle\dependency-check-data\7.0\cache
        Region [POM] Set maxKeySize to: '1000000'
        Region [POM] Loaded keys from [POM], key count: 105; up to 1000000 will be available.
        Region [POM] Finished inital consistency check, isOk = true in 0ms.
        Region [POM] Indexed Disk Cache is alive.
        
        [...]
        
        Parsed regions [CENTRAL, POM, NODEAUDIT]
        Finished configuration in 51 ms.

    - `CentralAnalyzer` now completes instantly due to 100% cache hits:

    

        [...]

        [org.apache.commons.jcs.engine.control.CompositeCache] get: key = https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom, localOnly = false
        [org.apache.commons.jcs.engine.memory.AbstractDoubleLinkedListMemoryCache] POM: getting item for key https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        
        [...]

        [org.apache.commons.jcs.engine.control.CompositeCache] Attempting to get from aux [POM] which is of type: DISK_CACHE
        [org.apache.commons.jcs.utils.struct.AbstractLRUMap] getting item  for key https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.apache.commons.jcs.utils.struct.AbstractLRUMap] LRUMap miss for https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.apache.commons.jcs.auxiliary.disk.indexed.IndexedDiskCache] Region [POM] Trying to get from disk: https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.apache.commons.jcs.utils.struct.AbstractLRUMap] getting item  for key https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.apache.commons.jcs.utils.struct.AbstractLRUMap] LRUMap hit for https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.apache.commons.jcs.auxiliary.disk.indexed.IndexedDiskCache] Region [POM] Found on disk, key: https://search.maven.org/remotecontent?filepath=org/projectlombok/lombok/1.18.10/lombok-1.18.10.pom
        [org.owasp.dependencycheck.xml.pom.Model@54e0245], attr [[ IS_LATERAL = false, IS_SPOOL = true, IS_REMOTE = false, IS_ETERNAL = true, MaxLifeSeconds = 7776000, IdleTime = -1, CreateTime = 1669576503525, LastAccessTime = 1669576503525, getTimeToLiveSeconds() = 7756128, createTime = 1669576503525 ]]
        [org.apache.commons.jcs.engine.control.CompositeCache] POM - Aux cache[POM] hit

        [...]

        Finished Central Analyzer (0 seconds)

### Suggested Fix

Ugrade to JCS3 Core. This solves [JCS issue 199](https://issues.apache.org/jira/browse/JCS-199) that was discovered in version 2.2.1 which is currently used by ODC.

Points to consider when moving from 2.x to 3.x are outlined on this [help page](https://commons.apache.org/proper/commons-jcs/UpgradingFrom2x.html). They raised the baseline to Java 8. The current ODC parent defines the same baseline.

In the MWE attached you may try out the Gradle plugin running with JCS core 3.1.
1. Delete cache data directory
2. Switch to `myproject`
3. Uncomment lines in `build.gradle` and `settings.gradle`
4. Comment line with plugin version 7.3.2 in `build.gradle`
5. Package JAR for ODC core project
6. `gradlew :dependencyCheckAggregate --info`
7. repeat step 6
    - same behavior as in point 5 from repro above but this time without having to deactivate the daemon

### LINKED ISSUE #5096: JCS log spam when using the Central Analyzer cache from Ant
**Describe the bug**
Dependency checking from Ant using the Central Analyzer cache outputs about a screen's worth of JCS logging at both the start and end of the output.

**Version of dependency-check used**
The problem occurs using version 7.3.1 of the Ant task.

**Log file**
See https://gist.github.com/slieschke/82434d9f030feeb3b4bc520527bb37ce for example output.

**To Reproduce**
Run a dependency check from Ant with `centralAnalyzerUseCache="true"`.

**Expected behavior**
See https://gist.github.com/slieschke/525e30902dc5cdad71f504e1630861a6 for expected output.

### PR DESCRIPTION
**Pull request by jeremylong**
Upgrade to JCS3 to resolve cache misses in the gradle implementation. Additionally, this allows easier control over the JCS logging - as the logs are now routed to slf4j.

This is a draft PR as we must wait until JCS 3.2 is released due to a [bug in the JCS logging](https://issues.apache.org/jira/browse/JCS-232).

Resolves https://github.com/jeremylong/DependencyCheck/issues/5096 
Resolves https://github.com/dependency-check/dependency-check-gradle/issues/293

### CODE REVIEW

**Comment by mroeppis** [comment]
[JCS 3.2](https://github.com/apache/commons-jcs/releases/tag/commons-jcs3-3.2) has been released 3 days ago.

@jeremylong : Will the upgrade make it into 8.0.0 milestone?

File: pom.xml [hunk 1]
```diff
@@ -147,7 +147,10 @@ Copyright (c) 2012 - Jeremy Long
         <commons-io.version>2.11.0</commons-io.version>
         <commons-lang3.version>3.12.0</commons-lang3.version>
         <commons-text.version>1.10.0</commons-text.version>
-        <commons-jcs-core.version>2.2.1</commons-jcs-core.version>
+        <!-- note that logging will be noisy and broken until we upgrade to 3.2
+            See https://issues.apache.org/jira/browse/JCS-232 and
+            https://github.com/apache/commons-jcs/pull/120 -->
+        <commons-jcs-core.version>3.1</commons-jcs-core.version>
```
Comment (jeremylong) [inline pom.xml]:
```suggestion
        <commons-jcs-core.version>3.2</commons-jcs-core.version>
```

**Comment by jeremylong** [comment]
3.2 does not appear to be in central yet (see [failing build](https://github.com/jeremylong/DependencyCheck/actions/runs/3901358029/jobs/6663107395))

**Comment by mroeppis** [comment]
My mistake. I was happy to see a stable version among tags. But as it [looks](https://github.com/apache/commons-jcs/tags) a release should have a `rel/` prefix. So the one from above was merely for preparation.

**Comment by jeremylong** [comment]
Asked about the release date here: https://issues.apache.org/jira/browse/JCS-233

**Comment by aikebah** [comment]
@jeremylong JCS 3.2 was released in September. Would you prefer a rebase or a merge for this PR to bring it up-to-date?

https://commons.apache.org/proper/commons-jcs/changes-report.html#a3.2

**Comment by jeremylong** [comment]
I just rebased this PR. I'll re-review what I did and likely merge this tomorrow (along with cutting the next release).

**Pull request merged**